### PR TITLE
Maps and sets

### PR DESCRIPTION
**Pull request by abassawo**

### CODE REVIEW

File: lessons/arrays-arraylists/README.md [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+So far, the reference to the array has been created, but the array itself is not yet created. In order to store and access elements, we must initialize the array.
+
+*initializing Array*
+Java provides two ways to initialize arrays.
+
+1) We can specify a length of the array using the following syntax:
+
+
+```java
+numbers = new int[10]
```
Comment (ekchang) [inline lessons/arrays-arraylists/README.md]: `int numbers = new int[10];`

File: lessons/arrays-arraylists/README.md [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+```
+The above code initializes numbers to an array with a length of 10.
+Note the key word new, which is used to create a new instance of object.
+
+Here is a visual interpretation of the array:
+
+[array-zeroes](array-of-zeroes.jpg)
```
Comment (ekchang) [inline lessons/arrays-arraylists/README.md]: You should be able to embed the image with:

array-zeroes

File: lessons/arrays-arraylists/README.md [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+Note: the numbering of elements starts from 0 All arrays in Java start with the zero index! Also, the default value of every element in an array of integers is 0 prior to setting the values of each element.
```
Comment (ekchang) [inline lessons/arrays-arraylists/README.md]: missing punctuation after first 0. I would rephrase "All arrays in Java start with zero index!" to be more like "Java is a zero indexed language." so it doesn't seem like only arrays are 0, but nearly every starting index for any collection is 0. Maybe even link to the wikipedia page as optional reading for the curious: https://en.wikipedia.org/wiki/Zero-based_numbering

File: lessons/arrays-arraylists/README.md [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+
+
+2) We can also directly initialize an array with desired elements by using the follow syntax:
+
+```java
+numbers = { 1, 2, 3, 4, 5};
+String[] names = {"Alice", "Bob", "Edsger", "Grace"}
```
Comment (ekchang) [inline lessons/arrays-arraylists/README.md]: I had to look up if Edsger was a real name and found out that it's Dijkstra's first name so A+ :+1: 

And numbers should still include the type declaration: `int[] numbers = { 1, 2, 3, 4, 5};` We shouldn't piggy back off the declaration made at the top of the page.

Missing semicolon at the end though. `String[] names = {"Alice", "Bob", "Edsger", "Grace"};` We must make sure every code snippet compiles or it will get really confusing really fast for students.

File: lessons/arrays-arraylists/README.md [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+```
+Both ways will create an array of fixed-size and with the same element type. The only difference is that the latter initialization approach defaults to selected values.
+
+Note that we can create arrays of any type. Above, we also have an array of Strings.
+
+
+3) *Accessing Arrays*
+Once your array has been declared and initialized, you can access any element by referencing the name of the array and the desired index. The only constraints are that the index number must be greater than or equal to zero (first index) and less than or equal to the array's length - 1 (last index).
+
+An array's .length property returns the numerical count of the array's
```
Comment (ekchang) [inline lessons/arrays-arraylists/README.md]: This sentence seems incomplete. Shouldn't we say "An array's `length` property returns the number of elements in the array"?

We should also give a brief example below:
```java
int[] numbers = new int[10];
System.out.println(numbers.length); // 10
```

File: lessons/arrays-arraylists/README.md [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+
+```java
+int firstNumberinArray = numbers[0];
```
Comment (ekchang) [inline lessons/arrays-arraylists/README.md]: <nit> I would like to flesh out these just a bit more, but maybe I'm being excessive. Something like:

```
numbers[0] = 4;
int firstNumberInArray = numbers[0];
System.out.println(firstNumberInArray); // Prints 4
```

But maybe I'm being excessive.

File: lessons/arrays-arraylists/README.md [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+```   
+
+We can also update any element in the array by assigning a new value to it.
+
+```java
+numbers[4] = 100;
```
Comment (ekchang) [inline lessons/arrays-arraylists/README.md]: \<nit> Same nit as above, something more expressive:

```
numbers[0] = 5;
System.out.println(numbers[0]); // Prints 5
numbers[0] = 50;
System.out.println(numbers[0]); // Prints 50
numbers[0] += 1;
System.out.println(numbers[0]); // Prints 51
```
Again, your call

File: lessons/arrays-arraylists/README.md [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+```
+
+
+***
+Important to remember
+- Arrays must have elements of the same type
```
Comment (ekchang) [inline lessons/arrays-arraylists/README.md]: I'd clarify this by adding "Your int[] arrays can't hold Strings and your char[] arrays can't hold doubles"

File: lessons/arrays-arraylists/README.md [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+Create a class `MovieListings` and follow along in it.
+
+
+Before we can use an `ArrayList`, we have to `import` it:
+
+```java
+import java.util.ArrayList;
+```
+
+Before looking at the constructor for an arraylist, here are some distinguishing features of an  arraylist:
```
Comment (ekchang) [inline lessons/arrays-arraylists/README.md]: Extra space before ArrayList, and capitalize whenever referring to the class name and L. I think it's important to be clear we are referring to a class/type and not a general concept (which is why I'm ok with lowercase list...until we start referring to the List interface)

File: lessons/arrays-arraylists/README.md [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+
+1) Arraylists do NOT have a fixed size. The size() method returns the number of elements the list contains.
+2) May contain elements of the same type OR elements of different types.
```
Comment (ekchang) [inline lessons/arrays-arraylists/README.md]: Wait, #2 is incorrect. The type arguments in the collection declaration imposes compile time safeties to guard against this. You can't declare:
```
List<String> names = new ArrayList<>();
names.add(4.0d); // crash: throws method Collection.add(String) is not applicable (argument mismatch; double cannot be converted to String)
```
While technically you can declare an ArrayList without the type declaration, you should never do it. If anything #2 should say you should treat the type the same as arrays: a list can only contain the same object as the declared type (by doing so in the angle brackets)

File: lessons/hashmaps-and-sets/exercises/ListMap.java [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+     * @param keys
+     *   The array of keys in the "map".
+     * @param values
+     *   The array of values in the "map".
+     * @param key
+     *   The key to add or replace in the map.
+     * @param value
+     *   The value associated with the key.
+     */
+    public static void put(String key, int value) {
```
Comment (ekchang) [inline lessons/hashmaps-and-sets/exercises/ListMap.java]: This will be pretty hard to do if it's `static`

Comment (ekchang) [inline lessons/hashmaps-and-sets/exercises/ListMap.java]: This will be pretty hard as a `static` method. In fact it is a bit misleading because `put` seems to imply we're going to use the instance variables `keys` and `values` declared above, but they're going to be in for a nasty surprise if they try to reference it...

File: lessons/hashmaps-and-sets/exercises/ListMap.java [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+     * @param keys
+     *   The array of keys in the "map".
+     * @param values
+     *   The array of values in the "map".
+     * @param key
+     *   The key to look up.
+     * @return
+     *   The value associated with the key, or null if the key is not in the "map".
+     */
+    public static Integer get(String key) {
```
Comment (ekchang) [inline lessons/hashmaps-and-sets/exercises/ListMap.java]: Same static issue above

File: lessons/hashmaps-and-sets/exercises/ListMap.java [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+     */
+    public static Integer remove(String key) {
+        // FIXME: Write this method!
+        return null;
+    }
+
+    public static void main(String[] args) {
+        // Create the empty keys and values columns.
+        ArrayList<String> keys = new ArrayList<String>();
+        ArrayList<Integer> values = new ArrayList<Integer>();
```
Comment (ekchang) [inline lessons/hashmaps-and-sets/exercises/ListMap.java]: This is a missed opportunity to just create a new `ListMap` instance:

```
ListMap listMap = new ListMap();
listMap.put("Neptune", 14);
...
```

File: lessons/hashmaps-and-sets/exercises/ListMap.java [hunk 14]
```diff
@@ -0,0 +1,98 @@
+public class ListMap {
+    private List<String> keys;
+    private List<Integer> values;
+
+    public ListMap(){
+        keys = new ArrayList();
+        values = new ArrayList();
```
Comment (ekchang) [inline lessons/hashmaps-and-sets/exercises/ListMap.java]: Collections should always be instantiated with their type parameter. Even though this compiles, it is unsafe. 
unchecked

We can touch upon it briefly above, but they should instantiate new collections with `new ArrayList<>();` where the compiler can infer the type parameter from the original declaration. I think it's better to teach good habits first, even if they don't know why, than to teach bad habits and try to correct them later.

File: lessons/arrays-arraylists/README.md [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+3) Common operations : add(), get(), remove(), size();
+
+Next, we call the constructor, but we have to declare the type of object the `ArrayList` is going to hold. This is how you create a new `ArrayList` holding `String` objects.
+
+```java
+ArrayList<String> names = new ArrayList<String>();
```
Comment (ekchang) [inline lessons/arrays-arraylists/README.md]: I thought about this for a bit, and I think we should immediately teach students to declare Collections using their interface rather than the concrete class. We won't get into interfaces, but to just get them used to the idea that "there are different kinds of lists, but they all have the same basic list methods" :

```
List<String> names = new ArrayList<String>();
```

I originally planned to let this slide, but I'm going to move forward with a mentality that it's better to develop good habits and not exactly know why you're doing it yet, rather than teach a bad habit and hope you can correct it later.

File: lessons/hashmaps-and-sets/exercises/SocialNetwork.java [hunk 16]
```diff
@@ -0,0 +1,68 @@
+import java.util.HashSet;
+
+public class SocialNetwork {
+
+    public static void main(String[] args) {
+        HashSet<String> yours = new HashSet();
```
Comment (ekchang) [inline lessons/hashmaps-and-sets/exercises/SocialNetwork.java]: Same here as above: should be `Set<String> yours = new HashSet<>();`
Same for `mine` below.

File: lessons/hashmaps-and-sets/exercises/maps-exercises.md [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+  }
+  ```
+
+  For example, if you call it with an array containing "hello" and "goodbye", you'll get back a hash map with two keys: "hello" associated with the value 5, and "goodbye" associated with the value 7.
+
+  Test it with code similar to this.
+
+  ```java
+
+  ArrayList<String> words = new ArrayList<String>();
```
Comment (ekchang) [inline lessons/hashmaps-and-sets/exercises/maps-exercises.md]: `List<String> words`

File: lessons/hashmaps-and-sets/exercises/maps-exercises.md [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+```
+
+Please work through these exercises, completing as many as you can in class.  
+
+1. **Finish moons:** Finish the HashMap of number of moons for each planet.
+
+1. **Moons as method:** Convert this to a _method_ that _returns_ this hashmap.  It should look like this:
+
+  ```java
+  public static HashMap<String, Integer> getNumberOfMoons() {
```
Comment (ekchang) [inline lessons/hashmaps-and-sets/exercises/maps-exercises.md]: return a `Map<String, Integer>`

File: lessons/hashmaps-and-sets/exercises/maps-exercises.md [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+  ```
+  (name) is not a state!
+  ```
+
+  Make sure you test your program on correct and incorrect state names.
+
+1. **Map of word lengths:** Write a method that takes an array list of strings as a parameter, and creates and returns a hash map whose keys are the strings from the array and whose values are the lengths of these strings.
+
+  ```java
+  public static HashMap<String, Integer> getLengths(ArrayList<String> strings) {
```
Comment (ekchang) [inline lessons/hashmaps-and-sets/exercises/maps-exercises.md]: `Map<String, Integer>`

File: lessons/hashmaps-and-sets/exercises/maps-exercises.md [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+  Test it with code similar to this.
+
+  ```java
+
+  ArrayList<String> words = new ArrayList<String>();
+  words.add("banana");
+  words.add("pineapple");
+  words.add("cantaloupe");
+
+  HashMap<String, Integer> lengths = getLengths(words);
```
Comment (ekchang) [inline lessons/hashmaps-and-sets/exercises/maps-exercises.md]: `Map<String, Integer>`

File: lessons/hashmaps-and-sets/exercises/maps-exercises.md [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+
+1. **(Bonus) Two lists make a map:**  Another way to represent a mapping is with the key and value columns in the tables
+
+shown above, using one `ArrayList` for the keys and one `ArrayList` for the columns.  
+
+  For example, to represent the number of moons for each planet, the keys will be a `ArrayList<String>` containing Mercury, Venus, Earth, Mars, _etc._ and the values will be a `ArrayList<Integer>` containing 0, 0, 1, 2, _etc._  To look up the value for a key, first find the index of the key in the keys list, and then extract the corresponding index from the values list.
+
+  The lesson project [`ListMap`](exercises/ListMap.java) contains a class named `ListMap`, which sets this up for you.  There are three methods for you to implement: one to put a key and value into the "map", one to look up a key, and one to remove a key and value.  The `main()` method contains tests: once you have implemented the other three methods correctly, all the tests should print "true"!
+
+  Note: While this is a correct way to implement maps in Java, the reason we don't usually do it this way is because it is inefficient once there are many keys and values in the map.
```
Comment (ekchang) [inline lessons/hashmaps-and-sets/exercises/maps-exercises.md]: I think this is a good opportunity to hint at a bonus exercise: Can you implement a Map with a single list? Maybe a type of class that holds 2 values?

File: lessons/hashmaps-and-sets/hashmap.md [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+
+A _map_ is a very useful collection that can be used to associate one object (the key) with another (the value). The key to the Map must be unique, and it will be used to retrieve the value at a later time. Java Map collections can only hold objects, and the Map interface specifies a strict contract about how to put entries into the Map and what types can be added.
+
+Here are some of the operations you'll do with maps
+- Put something in the Map
+- Get something out of the Map  
+- Get a Set of keys to the Map - for iterating over it.
+
+
+Because [Map] is an interface, you can't instantiate it directly. Java includes some concrete implementations of the Map interface.
```
Comment (ekchang) [inline lessons/hashmaps-and-sets/hashmap.md]: I like this whole introduction to Maps. We should do the same for Lists in the ArrayList intro.

File: lessons/hashmaps-and-sets/hashmap.md [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+-  The map can look things up only in one direction: given the key, return the value.  Because a value can appear more than once, it doesn't work the other way: for example, which planet has zero moons?
+
+# Java's HashMap
+
+There are several ways to make a map in Java. Today, we'll use `java.util.HashMap` ([API docs](http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html)). It implements the `Map` interface, and You'll have to import it like this, or let IntelliJ do it for you.
+
+Recall that with `ArrayList`, you have to tell Java the type of elements in angle brackets.  With `HashMap`, you have to tell Java the types of both the keys and the values.
+
+```java
+HashMap<String, Integer> numberOfMoons = new HashMap<String, Integer>();
```
Comment (ekchang) [inline lessons/hashmaps-and-sets/hashmap.md]: `Map<String, Integer> numberOfMoons`

File: lessons/hashmaps-and-sets/hashmap.md [hunk 24]
```diff
@@ -0,0 +1,118 @@
...
+```
+
+But any key other than one we've put in already returns null.
+
+```java
+System.out.println(numberOfMoons.get("Neptune"));
+System.out.println(numberOfMoons.get("Sun"));
+System.out.println(numberOfMoons.get("Texas"));
+```
+
```
Comment (ekchang) [inline lessons/hashmaps-and-sets/hashmap.md]: Don't forget to include the example of overwriting an existing value in the Map:
```
map.put("Earth", 1);
System.out.println(map.get("Earth")); // Prints 1
map.put("Earth", 10);
System.out.println(map.get("Earth")); // Prints 10

File: lessons/hashmaps-and-sets/sets.md [hunk 25]
```diff
@@ -0,0 +1,158 @@
...
+| Unique | No | Yes |
+| Ordered | Yes | Depends |
+| Indexable | Yes | No |
+
+HashMap's `keyset()` function returns a collection of unique keys in the Map.
+
+On occasion, you might find yourself with a reference to a Map, and you want to walk over its entire set of contents. In this case, you need a Set of the keys to the Map:
+
+```Java
+Set<String> keys = mapOfIntegers.keySet();
```
Comment (ekchang) [inline lessons/hashmaps-and-sets/sets.md]: haha surely you mean `setOfIntegers` right?

File: lessons/hashmaps-and-sets/hashmap.md [hunk 26]
```diff
@@ -0,0 +1,118 @@
...
+Here are some of the operations you'll do with maps
+- Put something in the Map
+- Get something out of the Map  
+- Get a Set of keys to the Map - for iterating over it.
+
+
+Because [Map] is an interface, you can't instantiate it directly. Java includes some concrete implementations of the Map interface.
+
+Today, we will learn about the HashMap, but first here are some examples of maps:
```
Comment (ekchang) [inline lessons/hashmaps-and-sets/hashmap.md]: We should be prepared to answer why it's called a `HashMap` and the same with `ArrayList`.

File: lessons/hashmaps-and-sets/sets.md [hunk 27]
```diff
@@ -0,0 +1,158 @@
...
+You import it like this:
+
+```java
+import java.util.HashSet;
+```
+
+let's start with an empty `HashSet`. Call it `odd`.
+
+```java
+HashSet<Integer> odd = new HashSet<Integer>();
```
Comment (ekchang) [inline lessons/hashmaps-and-sets/sets.md]: `Set<Integer> odd`

File: lessons/hashmaps-and-sets/sets.md [hunk 28]
```diff
@@ -0,0 +1,158 @@
...
+
+#### contains, size, isEmpty
+
+Sets also have our standard methods: `contains`, `size`, and `isEmpty`.
+
+```java
+odd.contains(1); // true
+```
+
+#### Cloning a set
```
Comment (ekchang) [inline lessons/hashmaps-and-sets/sets.md]: I...would be very careful about using the word `clone`. There is an explicit `Object#clone()` method and `Cloneable` interface and they are both very very hard to use correctly. Let's rename all references to `clone` as "Duplicating" or "Make a copy of a set"

File: lessons/hashmaps-and-sets/sets.md [hunk 29]
```diff
@@ -0,0 +1,158 @@
...
+
+An easy way to clone a set is to just pass the set you want to copy into the constructor:
+
+
+```java
+HashSet<Integer> odd2 = new HashSet<Integer>(odd);
+HashSet<Integer> odd3 = new HashSet<Integer>(odd);
```
Comment (ekchang) [inline lessons/hashmaps-and-sets/sets.md]: `Set<Integer> odd` for both

File: lessons/hashmaps-and-sets/sets.md (hunk 25 above)
Comment (abassawo) [inline lessons/hashmaps-and-sets/sets.md]: The mapOfIntegers this is referencing is a Map from the exercise that must map an int's String representation as its key, to its Integer Value. Not the most self-explanatory example. Could probably use a better name too. So, for example: calling map.get("1") for example would yield 1.

File: lessons/arrays-arraylists/README.md (hunk 10 above)
Comment (abassawo) [inline lessons/arrays-arraylists/README.md]: Thanks, Erick. #2 still stands. This is meant to be a brief side by side look at the Array and ArrayList objs. Of course, if you specify the type argument, then the entire point goes out of the window. I was hoping to provide brief overview of what the structure can do and what it's limitations are, and from there we can discuss what is considered good practice. Will incorporate these comments into next pr.
